13 split encryption from messages

HoltronDev opened the pull request:
Moved encryption out of the messages altogether. To encrypt will now be passing the message to the encryption algorithm.

HoltronDev committed:
Added AesGcm, made some modifications.

HoltronDev committed:
Fixing StringRead in NetBuffer.Read

StringRead() operated differently from StringRead(out string result). Made the logic from the working into the other and it seems to play nice now.

HoltronDev committed:
Removing old Encryption Stuff

instead of trying to get all the old encryption stuff working, I just opted to remove all of it. We'll rebuild new ones as we go along/need it. A method of using an interface for encryption has been added.

HoltronDev committed:
Update EncryptionTests.cs

joshkato commented on Holtron.Net/Network/NetBuffer.Read.cs:
```diff
@@ -564,41 +564,76 @@ public long ReadRangedInteger(long min, long max)
+            }
+
+            byte[] bytes;
+            if (ReadBytes((int)byteLen, out bytes) == false)
```
You can combine the `byte[] bytes` declaration with the `ReadBytes` call and drop the `== false` part.
```suggestion
            if (!ReadBytes((int)byteLen, out var bytes))
```

HoltronDev replied:
Done been changed.

joshkato commented on Holtron.Net/Network/NetBuffer.Read.cs:
```diff
@@ -564,41 +564,76 @@ public long ReadRangedInteger(long min, long max)
+
+            var result2 = Encoding.UTF8.GetString(bytes, 0, bytes.Length);
+            return result2;
+            //			if (byteLen <= 0)
```
All of this commented code and probably be removed.

HoltronDev replied:
Removed.

joshkato commented on Holtron.Net/Network/NetBuffer.cs:
```diff
@@ -94,5 +94,7 @@ static NetBuffer()
 			s_buffer = Array.Empty<NetBuffer>();
 		}
+
+        //public static implicit operator byte[](NetBuffer x) => x.m_data;
```
Remove unused code.

HoltronDev replied:
Removed.

HoltronDev committed:
Updated if, removed dead code

Removing unused/commented code and updating an if block.

joshkato commented on Holtron.Net/Network/Encryption/NetEncryptionAESGCM.cs:
```diff
@@ -0,0 +1,82 @@
+
+namespace Holtron.Net.Network.Encryption
+{
+    public class NetEncryptionAESGCM : INetEncryption
```
`NetEncryptionAESGCM` or `INetEncryption` should implement `IDisposable` since this is working with objects that should be disposed when the instance is released.

HoltronDev replied:
Implemented IDisposable on INetEncryption.

joshkato commented on Holtron.Net/Network/Encryption/NetEncryptionAESGCM.cs:
```diff
@@ -0,0 +1,82 @@
+
+        public NetEncryptionAESGCM(byte[] key)
+        {
+            var derivedKey = new Rfc2898DeriveBytes(key, new byte[8], 1000).GetBytes(16);
```
- `Rfc2898DeriveBytes` is disposable so its usage should probably be accompanied by a `using` unless you want to keep the instance of `Rfc2898DerivedBytes` around.
- Was using `{ 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }` as a salt intentional?
- I would move the number of rounds for PBKDF2 (`Rfc2898DeriveBytes`) into a class-level private `const`. You may also want to consider using more than 1000 rounds. It looks like OWASP recommends a significantly larger number of rounds: [Password Storage Cheat Sheet (PBKDF2)](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#pbkdf2).

HoltronDev replied:
I updated the SALT and encryption rounds. Changed it a little bit to allow for a using statement.

joshkato commented on Holtron.Net/Network/Encryption/NetEncryptionAESGCM.cs:
```diff
@@ -0,0 +1,82 @@
+            aesGcm = new AesGcm(derivedKey);
+        }
+
+        public NetEncryptionAESGCM(Span<byte> key)
```
You could probably chain this constructor into the one that takes `byte[]` for the key:

```csharp
public NetEncryptionAESGCM(Span<byte> key) :
    this(key.ToArray())
{
}
```

HoltronDev replied:
Chained it and another constructor to the first.

joshkato commented on Holtron.Net/Network/Encryption/NetEncryptionAESGCM.cs:
```diff
@@ -0,0 +1,82 @@
+
+            aesGcm.Decrypt(nonce, cipherText, tag, plainTextMessage);
+
+            message.Data = plainTextMessage.ToArray();
```
It looks like setting `message.Data` sets `message.m_data` under the hood. You probably only need one or the other.

HoltronDev replied:
Standardized using `message.m_data`.

joshkato commented on Holtron.Net/Network/Encryption/NetEncryptionAESGCM.cs:
```diff
@@ -0,0 +1,82 @@
+                ? stackalloc byte[encryptedDataLength]
+                : new byte[encryptedDataLength].AsSpan();
+
+            var nonce = encryptedData.Slice(0, 12);
```
Is the size of the `nonce` and `tag` always guaranteed to be 12 bytes? This might be variable depending on where the code is running.

HoltronDev replied:
The nonce and tag are somethin' you set as part of the implementation. In the case of this, I'm defaulting to the smallest size allowable per the AesGcm class constants. I am gonna go ahead and move these to constants though so we can change them more quickly if needed.

joshkato commented on Holtron.Net/Network/Encryption/NetEncryptionAESGCM.cs:
```diff
@@ -0,0 +1,82 @@
+
+        public void SetKey(byte[] data, int offset, int length)
+        {
+            
```
Might just want to stick a `throw new NotImplementedException()` in here just so it isn't accidentally used.

HoltronDev replied:
Done been stuck.

HoltronDev committed:
Multiple Changes

Added IDisposable to INetEncryption
Added constants to encryption
Using Message.m_data instead of Data
Throw NotImplementedException for Key Setting.
Added Salt Generation,
Increased rounds for key generation.
Added Const for encryption rounds.

HoltronDev committed:
Changed to using on driveBytes class

Added a using statement so it gets disposed properly.

HoltronDev committed:
Remove CONST

Removed some constant fields for AesGcm defined sizes.
